### PR TITLE
Rework the precision metric.

### PR DESCRIPTION
**Pull request by trivialfis**
I just realized this was not a documented feature before the PR.

- Rework the precision metric for both CPU and GPU.
- Mention it in the document.
- Clean up old support code for GPU ranking metric.

The new implementation supports only learning to rank. Please consider one of the existing library implementations like sklearn for classification tasks. The previous implementation claims support for binary classification in a comment, which is false.

It is extracted from https://github.com/dmlc/xgboost/pull/8822.

### CODE REVIEW

**Review by hcho3 — CHANGES_REQUESTED** [review]

File: src/metric/rank_metric.cc [hunk 1]
```diff
@@ -1,25 +1,6 @@
...
-// - The GPU metric types are *only* registered when xgboost is built for GPUs
-//
-// This is done for 2 reasons:
-// - Clear separation of CPU and GPU logic
-// - Sorting datasets containing large number of rows is (much) faster when parallel sort
-//   semantics is used on the CPU. The __gnu_parallel/concurrency primitives needed to perform
-//   this cannot be used when the translation unit is compiled using the 'nvcc' compiler (as the
-//   corresponding headers that brings in those function declaration can't be included with CUDA).
-//   This precludes the CPU and GPU logic to coexist inside a .cu file
-
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/metric/rank_metric.cc:22]: The GPU registry is removed in this PR, as the precision metric is the last one being rewritten.

File: src/common/ranking_utils.h [hunk 2]
```diff
@@ -372,12 +372,37 @@ bool IsBinaryRel(linalg::VectorView<float const> label, AllOf all_of) {
  *         both CPU and GPU.
  */
 template <typename AllOf>
-void CheckMapLabels(linalg::VectorView<float const> label, AllOf all_of) {
+void CheckPreLabels(StringView name, linalg::VectorView<float const> label, AllOf all_of) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/ranking_utils.h:375]: Done.

**Review by hcho3 — APPROVED** [review]

**Review by hcho3 — APPROVED** [review]
